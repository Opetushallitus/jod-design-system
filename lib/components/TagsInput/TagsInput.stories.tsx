import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import { TagsInput } from './TagsInput';

const meta = {
  title: 'TagsInput',
  component: TagsInput,
  tags: ['autodocs'],
} satisfies Meta<typeof TagsInput>;

export default meta;

type Story = StoryObj<typeof meta>;

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542-7646',
  },
};

export const Top: Story = {
  parameters: {
    ...parameters,
  },
  args: {
    label: 'Label text',
    placeholder: 'Lorem ipsum dolor sit amet',
    tags: ['dog', 'cat', 'fish', 'bird'],
    onValueChange: fn(),
  },
};

export const Bottom: Story = {
  parameters: {
    ...parameters,
  },
  args: {
    label: 'Label text',
    placeholder: 'Lorem ipsum dolor sit amet',
    tags: ['dog', 'cat', 'fish', 'bird'],
    onValueChange: fn(),
    inputPosition: 'bottom',
  },
};
