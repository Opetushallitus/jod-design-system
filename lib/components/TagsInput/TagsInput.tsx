import { useState } from 'react';
import { TagsInput as ArkTagsInput, TagsInputValueChangeDetails as ArkValueChangeDetails } from '@ark-ui/react';

export interface TagsInputProps {
  label: string;
  placeholder: string;
  tags: string[];
  onValueChange: (values: string[]) => void;
  inputPosition?: 'top' | 'bottom';
}

export const TagsInput = ({ label, placeholder, tags, onValueChange, inputPosition = 'top' }: TagsInputProps) => {
  const [selectedTag, setSelectedTag] = useState<string | null>(null);

  const highlightChange = (details: ArkTagsInput.HighlightChangeDetails) => {
    setSelectedTag(details.highlightedValue);
  };

  const onValueChangeHandler = (details: ArkValueChangeDetails) => {
    onValueChange(details.value);
  };

  return (
    <ArkTagsInput.Root
      className="flex flex-col"
      value={tags}
      onHighlightChange={highlightChange}
      onValueChange={onValueChangeHandler}
      validate={(details) => {
        return !details.value.includes(details.inputValue);
      }}
      onFocusOutside={() => setSelectedTag(null)}
      onBlur={() => setSelectedTag(null)}
    >
      <ArkTagsInput.Context>
        {(tagsInput) => (
          <>
            <ArkTagsInput.Label className="mb-4 text-form-label text-primary-gray">{label}</ArkTagsInput.Label>
            <div
              className={`${tagsInput.value.length > 0 ? 'gap-3' : ''} flex flex-col rounded-[10px] border-[5px] border-border-gray p-5`.trim()}
            >
              {inputPosition === 'top' && (
                <ArkTagsInput.Input placeholder={placeholder && `(${placeholder})`} className="w-full" />
              )}
              <ArkTagsInput.Control className="flex flex-wrap gap-3">
                {tagsInput.value.map((value, index) => (
                  <ArkTagsInput.Item
                    key={index}
                    index={index}
                    value={value}
                    className={`${selectedTag == value ? 'outline outline-2 outline-offset-[1.5px]' : ''} group inline-flex select-none items-center justify-center rounded-sm border-2 border-secondary-gray px-4 text-tag`.trim()}
                  >
                    <ArkTagsInput.ItemPreview>
                      <div className="inline-flex items-center gap-3">
                        <ArkTagsInput.ItemText className="hyphens-auto text-secondary-gray group-hover:underline">
                          {value}
                        </ArkTagsInput.ItemText>
                        <ArkTagsInput.ItemDeleteTrigger className="flex">
                          <span className="material-symbols-outlined size-24-bold h-fit select-none text-secondary-gray">
                            {'close'}
                          </span>
                        </ArkTagsInput.ItemDeleteTrigger>
                      </div>
                    </ArkTagsInput.ItemPreview>
                    <ArkTagsInput.ItemInput className="py-[5px]" />
                  </ArkTagsInput.Item>
                ))}
              </ArkTagsInput.Control>
              {inputPosition === 'bottom' && (
                <ArkTagsInput.Input placeholder={placeholder && `(${placeholder})`} className="w-full" />
              )}
            </div>
          </>
        )}
      </ArkTagsInput.Context>
    </ArkTagsInput.Root>
  );
};
