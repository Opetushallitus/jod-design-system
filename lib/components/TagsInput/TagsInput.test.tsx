import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { TagsInput } from './TagsInput';

describe('Snapshot testing', () => {
  it('matches the snapshot', () => {
    const { container } = render(
      <TagsInput label="Label here" placeholder="Placeholder" tags={['cat', 'bird', 'dog']} onValueChange={vi.fn()} />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('TagsInput', () => {
  it('renders the label', () => {
    const label = 'Test Label';
    render(<TagsInput label={label} placeholder="Placeholder" tags={['cat', 'bird', 'dog']} onValueChange={vi.fn()} />);
    const labelElement = screen.getByText(label);
    expect(labelElement).toBeInTheDocument();
  });

  it('renders the placeholder', () => {
    const placeholder = 'Test Placeholder';
    render(
      <TagsInput label="Test Label" placeholder={placeholder} tags={['cat', 'bird', 'dog']} onValueChange={vi.fn()} />,
    );
    const inputElement = screen.getByPlaceholderText(`(${placeholder})`);
    expect(inputElement).toBeInTheDocument();
  });
});
