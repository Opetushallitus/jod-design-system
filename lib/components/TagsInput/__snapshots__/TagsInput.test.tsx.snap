// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > matches the snapshot 1`] = `
<div>
  <div
    class="flex flex-col"
    data-part="root"
    data-scope="tags-input"
    dir="ltr"
    id="tags-input::r0:"
  >
    <label
      class="mb-4 text-form-label text-primary-gray"
      data-part="label"
      data-scope="tags-input"
      dir="ltr"
      for="tags-input::r0::input"
      id="tags-input::r0::label"
    >
      Label here
    </label>
    <div
      class="gap-3 flex flex-col rounded-[10px] border-[5px] border-border-gray p-5"
    >
      <input
        autocapitalize="none"
        autocomplete="off"
        autocorrect="off"
        class="w-full"
        data-part="input"
        data-scope="tags-input"
        dir="ltr"
        id="tags-input::r0::input"
        placeholder="(Placeholder)"
        value=""
      />
      <div
        class="flex flex-wrap gap-3"
        data-part="control"
        data-scope="tags-input"
        dir="ltr"
        id="tags-input::r0::control"
      >
        <div
          class="group inline-flex select-none items-center justify-center rounded-sm border-2 border-secondary-gray px-4 text-tag"
          data-part="item"
          data-scope="tags-input"
          data-value="cat"
          dir="ltr"
        >
          <div
            data-part="item-preview"
            data-scope="tags-input"
            data-value="cat"
            dir="ltr"
            id="tags-input::r0::tag:cat:0"
          >
            <div
              class="inline-flex items-center gap-3"
            >
              <span
                class="hyphens-auto text-secondary-gray group-hover:underline"
                data-part="item-text"
                data-scope="tags-input"
                dir="ltr"
              >
                cat
              </span>
              <button
                aria-label="Delete tag cat"
                class="flex"
                data-part="item-delete-trigger"
                data-scope="tags-input"
                dir="ltr"
                id="tags-input::r0::tag:cat:0:delete-btn"
                tabindex="-1"
                type="button"
              >
                <span
                  class="material-symbols-outlined size-24-bold h-fit select-none text-secondary-gray"
                >
                  close
                </span>
              </button>
            </div>
          </div>
          <input
            aria-hidden="true"
            aria-label="Editing tag cat. Press enter to save or escape to cancel."
            class="py-[5px]"
            data-part="item-input"
            data-scope="tags-input"
            dir="ltr"
            hidden=""
            id="tags-input::r0::tag:cat:0:input"
            tabindex="-1"
            value=""
          />
        </div>
        <div
          class="group inline-flex select-none items-center justify-center rounded-sm border-2 border-secondary-gray px-4 text-tag"
          data-part="item"
          data-scope="tags-input"
          data-value="bird"
          dir="ltr"
        >
          <div
            data-part="item-preview"
            data-scope="tags-input"
            data-value="bird"
            dir="ltr"
            id="tags-input::r0::tag:bird:1"
          >
            <div
              class="inline-flex items-center gap-3"
            >
              <span
                class="hyphens-auto text-secondary-gray group-hover:underline"
                data-part="item-text"
                data-scope="tags-input"
                dir="ltr"
              >
                bird
              </span>
              <button
                aria-label="Delete tag bird"
                class="flex"
                data-part="item-delete-trigger"
                data-scope="tags-input"
                dir="ltr"
                id="tags-input::r0::tag:bird:1:delete-btn"
                tabindex="-1"
                type="button"
              >
                <span
                  class="material-symbols-outlined size-24-bold h-fit select-none text-secondary-gray"
                >
                  close
                </span>
              </button>
            </div>
          </div>
          <input
            aria-hidden="true"
            aria-label="Editing tag bird. Press enter to save or escape to cancel."
            class="py-[5px]"
            data-part="item-input"
            data-scope="tags-input"
            dir="ltr"
            hidden=""
            id="tags-input::r0::tag:bird:1:input"
            tabindex="-1"
            value=""
          />
        </div>
        <div
          class="group inline-flex select-none items-center justify-center rounded-sm border-2 border-secondary-gray px-4 text-tag"
          data-part="item"
          data-scope="tags-input"
          data-value="dog"
          dir="ltr"
        >
          <div
            data-part="item-preview"
            data-scope="tags-input"
            data-value="dog"
            dir="ltr"
            id="tags-input::r0::tag:dog:2"
          >
            <div
              class="inline-flex items-center gap-3"
            >
              <span
                class="hyphens-auto text-secondary-gray group-hover:underline"
                data-part="item-text"
                data-scope="tags-input"
                dir="ltr"
              >
                dog
              </span>
              <button
                aria-label="Delete tag dog"
                class="flex"
                data-part="item-delete-trigger"
                data-scope="tags-input"
                dir="ltr"
                id="tags-input::r0::tag:dog:2:delete-btn"
                tabindex="-1"
                type="button"
              >
                <span
                  class="material-symbols-outlined size-24-bold h-fit select-none text-secondary-gray"
                >
                  close
                </span>
              </button>
            </div>
          </div>
          <input
            aria-hidden="true"
            aria-label="Editing tag dog. Press enter to save or escape to cancel."
            class="py-[5px]"
            data-part="item-input"
            data-scope="tags-input"
            dir="ltr"
            hidden=""
            id="tags-input::r0::tag:dog:2:input"
            tabindex="-1"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
